lib/instrumentation/span-compression.js

```diff
@@ -0,0 +1,176 @@
+  // @param Span toCompressSpan
+  // @returns boolean|String
+  _getCompressionStrategy (compositeSpan, toCompressSpan) {
+    if (!this._isEnabled() || !compositeSpan._destination || !toCompressSpan._destination) {
+      return false
+    }
+
+    const isSameKind = compositeSpan.type === toCompressSpan.type &&
+    compositeSpan.subtype === toCompressSpan.subtype &&
+    compositeSpan._destination.service.resource === toCompressSpan._destination.service.resource
+
+    if (!isSameKind) {
+      return false
+    }
+
+    let strategy = STRATEGY_SAME_KIND
+    if (compositeSpan.name === toCompressSpan.name) {
+      strategy = STRATEGY_EXACT_MATCH
+    }
+
+    // check duration limits
+    if (!this._checkDuration(compositeSpan._duration, toCompressSpan._duration, strategy)) {
+      return false
+    }
+
+    return strategy
+  }
+
+  // Sets initial composite values or confirms strategy matches
+  //
+  // Returns true if spanToCompress can be compressed into compositeSpan,
+  // returns false otherwise.
+  //
+  // @param Span compositeSpan
+  // @param Span spanToCompress
+  // @returns boolean
+  _initCompressionStrategy (compositeSpan, spanToCompress) {
+    if (!this.composite.compression_strategy) {
+      // If no strategy is set, check if strategizable or not. If so,
+      // set initial values.  If not, bail.
+      this.composite.compression_strategy = this._getCompressionStrategy(
+        compositeSpan,
+        spanToCompress
+      )
+      if (!this.composite.compression_strategy) {
+        return false
+      }
+
+      // set initial composite context values
+      this.timestamp = compositeSpan.timestamp
+      this.composite.count = 1
+      this.composite.sum = compositeSpan._duration
+    } else {
+      // if so, compare with the compression strat and bail if mismatch
+      const strat = this._getCompressionStrategy(compositeSpan, spanToCompress)
+      if (strat !== this.composite.compression_strategy) {
+        return false
+      }
+    }
+
+    return true
+  }
+
+  // Helper method for duration checking
+  //
+  // Checks whether the additional time will "fit" within the
+  // current duration based on a span compression strategy
+  // constant.
+  //
+  // @param int currentMs
+  // @param int additionalMs
+  // @param String type
+  // @returns boolean
+  _checkDuration (currentMs, additionalMs, type) {
+    const totalDuractionAsSeconds = (currentMs + additionalMs) / 1000
+
+    if (type === STRATEGY_EXACT_MATCH && (totalDuractionAsSeconds < this._agent._conf.spanCompressionExactMatchMaxDuration)) {
+      return true
+    }
+
+    if (type === STRATEGY_SAME_KIND && (totalDuractionAsSeconds < this._agent._conf.spanCompressionSameKindMaxDuration)) {
+      return true
+    }
+
+    return false
+  }
+
+  // Attempts to compression the second span into the first
+  //
+  // @param Span compositeSpan
+  // @param Span spanToCompress
+  // @return boolean
+  tryToCompress (compositeSpan, spanToCompress) {
```

Comment:
The `tryToCompress` implementation would be a lot easier to compare to the spec if it followed the pseudo-code at https://github.com/elastic/apm/blob/main/specs/agents/handling-huge-traces/tracing-spans-compress.md#turning-compressed-spans-into-a-composite-span   Is there a particular reason to not have followed that?

Reply:
That's interesting -- I generally don't implement pseudo code literally from specs and  the couple of other agent's I looked at drifted a bit from it as well.  I'm not really sure how to answer that question other than I simplified the implementation details of that pseudo code and created the compression service object because 1. it made it easier to reason about and test the end result as the feature was being developed and 2. because I thought the complexity might be objected to during review.  

That said with the end-to-end tests in place refactoring this shouldn't be a big issue other than the time it will take, so I'll get started on that.